pynamodb/attributes.py

```diff
@@ -257,3 +261,223 @@ def deserialize(self, value):
         Takes a UTC datetime string and returns a datetime object
         """
         return parse(value, dayfirst=False).datetime
+
+
+class MapAttributeMeta(type):
+    def __init__(cls, name, bases, attrs):
+        setattr(cls, '_attributes', None)
+
+
+class MapAttribute(with_metaclass(MapAttributeMeta, Attribute)):
+    _attributes = None
+
+    def __init__(self, **attrs):
+
+        hash_key = attrs.get('hash_key', False)
+        range_key = attrs.get('range_key', False)
+        null = attrs.get('null', None)
+        default = attrs.get('default', None)
+        attr_name = attrs.get('attr_name', None)
+
+        super(MapAttribute, self).__init__(hash_key=hash_key,
+                                           range_key=range_key,
+                                           null=null,
+                                           default=default,
+                                           attr_name=attr_name)
+        self.attribute_values = {}
+        self._set_attributes(**attrs)
+
+    def __iter__(self):
+        return iter(self.attribute_values)
+
+    def __getitem__(self, item):
+        return self.attribute_values[item]
+
+    @classmethod
+    def _get_attributes(cls):
+        """
+        Returns the list of attributes for this class
+        """
+        if cls._attributes is None:
+            cls._attributes = AttributeDict()
+            for item in dir(cls):
+                try:
+                    item_cls = getattr(getattr(cls, item), "__class__", None)
+                except AttributeError:
+                    continue
+                if item_cls is None:
+                    continue
+                if issubclass(item_cls, (Attribute,)):
+                    instance = getattr(cls, item)
+                    if instance.attr_name is None:
+                        instance.attr_name = item
+                    cls._attributes[item] = instance
+        return cls._attributes
+
+    def _set_attributes(self, **attrs):
+        """
+        Sets the attributes for this object
+        """
+        for attr_name, attr in self._get_attributes().aliased_attrs():
+            if attr.attr_name in attrs:
+                if type(attrs.get(attr.attr_name)) is type(attr) or not isinstance(attrs.get(attr_name), collections.Mapping):
+                    setattr(self, attr_name, attrs.get(attr.attr_name))
+                else:
+                    thing = attrs.get(attr_name)
+                    instance = type(attr)()
+                    aliased_attrs = type(attr)._get_attributes().aliased_attrs()
+                    for bad_k,v in thing.iteritems():
```

Comment:
This for block (along with `bad_k` and `good_k`) is pretty confusing. Can you explain with comments what's going on and use different variable names?